### PR TITLE
Reduce duplication in construction of clients

### PR DESCRIPTION
**Pull request by nictas**

### CODE REVIEW

File: cloudfoundry-client-lib/src/main/java/org/cloudfoundry/client/lib/rest/CloudControllerRestClientImpl.java [hunk 1]
```diff
@@ -573,18 +573,14 @@ public CloudApplication getApplication(String applicationName, boolean required)
...
-    @Override
-    public CloudApplication getApplication(UUID applicationGuid, boolean required) {
         Map<String, Object> resource = findApplicationResource(applicationGuid);
         CloudApplication application = null;
         if (resource != null) {
             application = mapCloudApplication(resource);
         }
-        if (application == null && required) {
-            throw new CloudOperationException(HttpStatus.NOT_FOUND, "Not Found", "Application " + applicationGuid + " not found.");
+        if (application == null) {
```
Comment (enchobelezirev) [inline cloudfoundry-client-lib/src/main/java/org/cloudfoundry/client/lib/rest/CloudControllerRestClientImpl.java]: By default required?

File: cloudfoundry-client-lib/src/main/java/org/cloudfoundry/client/lib/rest/CloudControllerRestClientImpl.java [hunk 2]
```diff
@@ -1031,23 +1027,42 @@ public CloudServiceInstance getServiceInstance(String serviceName, boolean requi
...
 
     @Override
-    public CloudSpace getSpace(String spaceName) {
-        return getSpace(spaceName, true);
-    }
-
-    @Override
-    public CloudSpace getSpace(String spaceName, boolean required) {
-        Map<String, Object> resource = findSpaceResource(spaceName);
+    public CloudSpace getSpace(UUID spaceGuid) {
```
Comment (enchobelezirev) [inline cloudfoundry-client-lib/src/main/java/org/cloudfoundry/client/lib/rest/CloudControllerRestClientImpl.java]: Same here? Are there some methods which are hoping no exception to be thrown here?

File: cloudfoundry-client-lib/src/main/java/org/cloudfoundry/client/lib/rest/CloudControllerRestClientFactory.java [hunk 3]
```diff
@@ -80,31 +85,26 @@ public CloudControllerRestClient createClient(URL controllerUrl, CloudCredential
...
+        return createClient(controllerUrl, credentials, organizationName, spaceName, createOAuthClient(controllerUrl));
     }
 
     public CloudControllerRestClient createClient(URL controllerUrl, CloudCredentials credentials, String organizationName,
         String spaceName, OAuthClient oAuthClient) {
-        this.oAuthClient = oAuthClient;
-        LoggregatorClient loggregatorClient = new LoggregatorClient(trustSelfSignedCerts);
+        CloudControllerRestClientImpl clientWithoutTarget = new CloudControllerRestClientImpl(controllerUrl, credentials, restTemplate,
+            oAuthClient, null, null);
+        CloudSpace target = clientWithoutTarget.getSpace(organizationName, spaceName);
```
Comment (enchobelezirev) [inline cloudfoundry-client-lib/src/main/java/org/cloudfoundry/client/lib/rest/CloudControllerRestClientFactory.java]: clientWithoutTarget can retrieve space o.O

Comment (nictas) [reply cloudfoundry-client-lib/src/main/java/org/cloudfoundry/client/lib/rest/CloudControllerRestClientFactory.java]: Yup, you can do `cf spaces` and `cf orgs` even if you don't have a target.

File: cloudfoundry-client-lib/src/main/java/org/cloudfoundry/client/lib/rest/CloudControllerRestClientImpl.java (hunk 2 above)
Comment (nictas) [reply cloudfoundry-client-lib/src/main/java/org/cloudfoundry/client/lib/rest/CloudControllerRestClientImpl.java]: We do `restTemplate.getForObject()`, which throws an exception if the backend returns 404. As a result, this check is unreachable and essentially dead code.

File: cloudfoundry-client-lib/src/main/java/org/cloudfoundry/client/lib/rest/CloudControllerRestClientImpl.java (hunk 1 above)
Comment (nictas) [reply cloudfoundry-client-lib/src/main/java/org/cloudfoundry/client/lib/rest/CloudControllerRestClientImpl.java]: Same as in `getSpace`.

**Pull request merged**